File: sem3/matr/e.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    freopen("cycles.in", "r", stdin);
    freopen("cycles.out", "w", stdout);

    long long n, m;
    cin >> n >> m;
    vector<long long> w(n);
    for (long long i = 0; i < n; ++i) {
        cin >> w[i];
    }
    set<long long> s;
    for (long long i = 0; i < m; ++i) {
        long long a, b, c = 0;
        cin >> a;
        for (long long j = 0; j < a; ++j) {
            cin >> b;
            c |= 1 << (b - 1);
        }
        s.insert(c);
    }
    vector<long long> q(s.begin(), s.end());
    while (!q.empty()) {
        long long i = q.back();
        q.pop_back();
        for (long long j = 0; j < n; ++j) {
            long long k = i | (1 << j);
            if (s.count(k) == 0) {
                q.push_back(k);
                s.insert(k);
            }
        }
    }
    long long t = 0, r = 0;
    for (long long i = (1 << n) - 1; i >= 0; --i) {
        if (s.count(i) == 0) {
            long long z = 0, c = 0, j = 0;
            for (long long k = i; k > 0; k >>= 1, ++j) {
                if (k & 1) {
                    z += w[j];
                    ++c;
                }
            }
            if (c > t) {
                t = c;
                r = z;
            } else if (c == t) {
                r = max(r, z);
            }
        }
    }
    cout << r;
}

File: sem3/matr/e.py
import sys, queue

sys.stdin = open('cycles.in', 'r')
sys.stdout = open('cycles.out', 'w')

n, m = map(int, input().split())
w = list(map(int, input().split()))
s = set()
for _ in range(m):
    a = list(map(int, input().split()[1:]))
    b = 0
    for i in a:
        b += 1 << (i - 1)
    s.add(b)

q = list(s)
while q:
    i = q.pop()
    for j in range(n):
        k = i | (1 << j)
        if not k in s:
            q.append(k)
            s.add(k)
t = 0
r = 0
for i in range((1 << n) - 1, -1, -1):
    if not i in s:
        z = 0
        for j in range(n):
            

        # if bin(i).count('1') > z:
        #     z = bin(i).count('1')
        #     r = sum(map(lambda x: w[x] * (bin(i)[len(bin(i)) - x - 1] == '1'), range(len(bin(i)) - 2)))
        # else:
        #     r = max(r, sum(map(lambda x: w[x] * (bin(i)[len(bin(i)) - x - 1] == '1'), range(len(bin(i)) - 2))))
print(r)
